cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java

```diff
@@ -180,58 +182,94 @@ public void run() {
+        _executorState = executorStateSupplier.get();
+        isExpectedState = true;
+      }
+
+      if (isExpectedState) {
+        moveRunnable.run();
+      }
+    }
+
+    /**
+     * Start the actual execution of the proposals in order: First move replicas, then transfer leadership.
      */
     private void execute() {
       _zkUtils = ZkUtils.apply(_zkConnect, 30000, 30000, false);
       try {
-        if (_state.compareAndSet(ExecutorState.State.EXECUTION_STARTED,
-                                 ExecutorState.State.REPLICA_MOVEMENT_TASK_IN_PROGRESS)) {
-          moveReplicas();
-        }
-        // Start leader movements.
-        if (_state.compareAndSet(ExecutorState.State.REPLICA_MOVEMENT_TASK_IN_PROGRESS,
-                                 ExecutorState.State.LEADER_MOVEMENT_TASK_IN_PROGRESS)) {
-          moveLeaders();
-        }
+        // 1. Move replicas if possible.
+        maybeMove(ExecutorState.State.EXECUTION_STARTED,
+                  ExecutorState.State.REPLICA_MOVEMENT_TASK_IN_PROGRESS,
+                  this::moveReplicas,
+                  () -> ExecutorState.replicaMovementInProgress(_numFinishedPartitionMovements,
+                                                                _executionTaskManager.getExecutionTasksSummary(),
+                                                                _finishedDataMovementInMB));
+
+        // 2. Transfer leadership if possible.
+        maybeMove(ExecutorState.State.REPLICA_MOVEMENT_TASK_IN_PROGRESS,
+                  ExecutorState.State.LEADER_MOVEMENT_TASK_IN_PROGRESS,
+                  this::moveLeaderships, ExecutorState::leaderMovementInProgress);
+
       } catch (Throwable t) {
         LOG.error("Executor got exception during execution", t);
       } finally {
-        // Add the null pointer check for unit test.
-        if (_loadMonitor != null) {
-          _loadMonitor.resumeMetricSampling();
-        }
-        _stopRequested = false;
+        _loadMonitor.resumeMetricSampling();
+        _stopRequested.set(false);
         _executionTaskManager.clear();
         KafkaCruiseControlUtils.closeZkUtilsWithTimeout(_zkUtils, 10000);
-        _state.set(ExecutorState.State.NO_TASK_IN_PROGRESS);
-        _zkUtils.close();
+        _state = ExecutorState.State.NO_TASK_IN_PROGRESS;
+        // It is possible that the _executorState might be outdated if the user checks it between the two assignments.
+        _executorState = ExecutorState.noTaskInProgress();
       }
     }
 
+    /**
+     * Get the in execution data to move in MB.
+     * @param inExecutionTasks Tasks that are either in progress or aborting.
+     * @return the in execution data to move in MB.
+     */
+    private long inExecutionDataToMoveInMB(Set<ExecutionTask> inExecutionTasks) {
+      return inExecutionTasks.stream().mapToLong(task -> task.proposal().dataToMoveInMB()).sum();
+    }
+
     private void moveReplicas() {
```

Comment:
I did not see the `_executorState` get updated when move the partitions. Am I missing something?

Reply:
Executor state is updated in `maybeMove()`.

Reply:
But during the movement, the executor state may change, right? i.e. the in-progress tasks, completed tasks, etc.